### PR TITLE
add plugin for checking broken links

### PR DESCRIPTION
**Pull request by iacore**
it works on my end
quality not guaranteed

### CODE REVIEW

**Comment by iacore** [comment]
it works with resolve_urls
nice!

File: check_links/mod.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ */
+export default function (userOptions?: Options) {
+  const options = merge(defaults, userOptions);
+
+  return (site: Site) => {
+    site.use(modifyUrls({
+      fn(url, page, _element) {
+        const url_base = site.options.location;
+        const urls: string[] = [];
+        for (const page of site.pages) {
```
Comment (oscarotero) [inline check_links/mod.ts:25]: This loop is executed for every url found in the html code. It should be moved outside this function and run only once. For example:

```js
const urls = new Set<string>(); // Set is more performant than arrays

site.process("*", (pages) => {
  url.clear(); // Clear on rebuild
  for (const page of pages) {
    urls.add(site.url(page.data.url, true)); // site.url() return the full url if the second argument is true
  }
  for (const file of site.files) {
    urls.add(site.url(file.outputPath, true));
  }
});

site.use(modifyUrls({
  // Use the urls
})
```

**Review by oscarotero — CHANGES_REQUESTED** [review]

**Comment by iacore** [comment]
should still merge modify_urls.ts into this

**Comment by oscarotero** [comment]
nice! Can I merge it?

**Comment by iacore** [comment]
new, new finding! Add this in document:

this plugin must be put after resolve_urls and relative_urls, or it will report false broken links. ideally, put it as the last plugin.


> nice! Can I merge it?

you can merge it now.

**Comment by iacore** [comment]
more ux

this plugin is great. while muffet will turn a blind eye to `/a` -> `/a/`, this plugin will not.

it also checks if files by `site.copy` is... i guess i found another bug?


the issue is, let's say i have a file `a.md` that is referenced in `index.md` as `[a](./a.md)`.

i have resolve_urls turned on.

resolve_urls will rewrite the url to `/a/index.html` (wtf?)

**Comment by oscarotero** [comment]
resolve_urls doesn't replace urls for pages that don't exist.
If you copy a.md, you should have a.md in your dest folder.

**Comment by oscarotero** [comment]
> the issue is, let's say i have a file a.md that is referenced in index.md as [a](./a.md).

Okay, I just reproduced this issue. It's a bug, will fix it.

**Comment by oscarotero** [comment]
Sorry, I merged this PR without review the code (I thought you only made my previous suggestions).
Why did you copy the code from modify_urls to this plugin?

**Comment by iacore** [comment]
> Why did you copy the code from modify_urls to this plugin?

modify_urls uses `node.setAttribute` while this plugin doesn't. might be faster.

**Comment by oscarotero** [comment]
Ah, okay.
If this plugin is merged in Lume repo, maybe it's a good idea to move this logic to a external function shared by both plugins to remove duplicated code.

**Comment by iacore** [comment]
> Ah, okay. If this plugin is merged in Lume repo, maybe it's a good idea to move this logic to a external function shared by both plugins to remove duplicated code.

maybe you can add a comment at the top of the file for this.

**Pull request merged**